### PR TITLE
feat: Deployments page

### LINKED ISSUE #4366: Kubernetes Deployments page
### Is your feature request related to a problem? Please describe

Deployments are the main way workloads are provisioned on Kubernetes, but we have no way to view them.

### Describe the solution you'd like

Add a new main nav item for Kubernetes Deployments, a list page, and a details page - similar to what we have for Containers or Pods. Users should be able to delete deployments.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by deboer-tim**
### What does this PR do?

Initial Deployments page. Creates a new route and icon in the main nav, plus
a regular page/table with deployment details. All columns from the design are
included, along with delete action and multiple deletion.

Notes:
- Deployments are not refreshing. It is likely unrelated to this PR, but
  should be resolved before this PR is merged. If you have existing
  deployments or Force Refresh everything else is working. (Fixed via #4989)
- In the design the Pods column are dots like the Pod list containers. This
  is deferred to a separate PR.
- We could sort by more columns... but not critical/can be done later.
- We should hide the nav item when there are no active Kubernetes providers. (PR updated to hide Deployments until you have a Kubernetes context)

### Screenshot/screencast of this PR

Screenshot 2023-11-23 at 3 16 18 PM

### What issues does this PR fix or reference?

Final major part of #4366.

### How to test this PR?

Set kube context to a cluster with a deployment.

### CODE REVIEW

**Comment by deboer-tim** [comment]
Pushed new branch based on main with the final changes, ready for regular review.

**Comment by lstocchi** [comment]
Played with it and it looks nice. I'll give my LGTM after we have https://github.com/containers/podman-desktop/issues/4974 otherwise it looks like it doesn't work. If you delete a deploy the table does not get refreshed, ... I'll give it a look for the ingress part

**Comment by lstocchi** [comment]
Other thing is, should we display some message if there are no deployments? This is what i see now
image

**Comment by deboer-tim** [comment]
Added new commit. Here's my view on where things stand:
- We should not merge until there is a fix for #4974. I've put up PR #4989 that solves the problem for deployments, but we need to merge that or a broader solution.
- The new commit hides the Deployment icon from the main nav when there are no Kubernetes contexts. I think this is a reasonable simple test for now, since we want to keep the UI simple when you aren't using Kubernetes, and when you are you could have a provider within Podman Desktop but could just have created a context externally. 
- The commit adds the missing empty screens for no deployments or all deployments filtered by search. There is no need
for a 'no provider' screen like other list pages because of the nav hiding.
- The commit also adds tests that were missing - basic UI and empty screens.

File: packages/renderer/src/lib/deployments/DeploymentColumnNamespace.svelte
Comment (jeffmaury) [inline packages/renderer/src/lib/deployments/DeploymentColumnNamespace.svelte]: I'm wondering why we don't have a generic NamespaceColumn as it's likely that we will have a lot of Kubernetes objects that are namespace aware

Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/DeploymentColumnNamespace.svelte]: I meant to open an issue after your comment in the Volume (?) PR, and did now: #5002. I was not trying to support reusing/sharing columns in the first pass of the table component, and right now they're still sort of typed to the object. I think this a very logical thing to do (soon), but the code/complexity reduction is fairly small so I have been focusing on getting at least two tables merged so that we 'have this problem' first.

File: packages/renderer/src/lib/deployments/DeploymentUI.ts
Comment (benoitf) [inline packages/renderer/src/lib/deployments/DeploymentUI.ts]: I think you need to review other places where we need to use the optional `?` vs `| undefined`
```
message? : string
```

File: packages/renderer/src/lib/deployments/DeploymentUI.ts
Comment (benoitf) [inline packages/renderer/src/lib/deployments/DeploymentUI.ts]:
```suggestion
  created?: Date;
```

File: packages/renderer/src/lib/deployments/DeploymentUI.ts
Comment (benoitf) [inline packages/renderer/src/lib/deployments/DeploymentUI.ts]:
```suggestion
  message?: string;
```

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (benoitf) [inline packages/renderer/src/lib/deployments/deployment-utils.ts]: should be simplified using map like approach or reduce like approach
```suggestion
conditions = (deployment.status.conditions ?? []).map(c => {
          type: con.type,
          message: con.message,
        }
);
```

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (benoitf) [inline packages/renderer/src/lib/deployments/deployment-utils.ts]: can use maybe the same approach then previous lines with ?? or || if there is no value
or we let humanizeAge accept undefined and it returns '' in that case

File: packages/renderer/src/lib/deployments/DeploymentUI.ts
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/DeploymentUI.ts]: Sorry, one of the drawbacks of having multiple branches and PR reviews atm is that I thought this had been merged as part of the original deployment store and I'd go back and fix it later based on the service store review. 🤦🏼‍♂️ Fixed via new commit.

File: packages/renderer/src/lib/deployments/DeploymentUI.ts
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/DeploymentUI.ts]: Fixed in new commit.

File: packages/renderer/src/lib/deployments/DeploymentUI.ts
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/DeploymentUI.ts]: Fixed in new commit.

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/deployment-utils.ts]: Fixed in new commit.

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/deployment-utils.ts]: I tried a couple ways and went with humanizeAge() accepting undefined since it seemed like the shortest/simplest code to me.

**Comment by deboer-tim** [comment]
I think I've addressed all the issues raised, but this PR is stagnating. Could I please get another review or comment on what else this needs?

**Review by cdrage — APPROVED** [review]
No issues with the code, this looks absolutely amazing.

The only feedback that I have is with the "Pods" column, I expect that if I click on the numbers, I'm taking to the Pods screen / be shown which Pods may be unavailable / not working / etc.

LGTM though, but maybe we should open up an issue to improve that column / UI design.

**Review by cdrage — CHANGES_REQUESTED** [review]
After my review, I realized while I was playing more with it that there is no sorting for Pods or Age (I understand why we cant sort Conditions)?

Screenshot 2023-11-30 at 10 32 49 PM

**Comment by deboer-tim** [comment]
Sorry, I should have explained - I was worried about both the size of this PR and it dragging on, so I was trying to find a reasonable spot to stop and why it says "Final major part of 4366" and not "Fixes".

You've listed exactly what I still planned to do: we should have pod linking and column sorting, and the design has the pods column like the pod view's container dots. If you think this isn't a good place to merge I can withdraw and include more of that here.

File: packages/renderer/src/lib/deployments/DeploymentColumnAge.svelte
Comment (benoitf) [inline packages/renderer/src/lib/deployments/DeploymentColumnAge.svelte]: now it seems you can use your custom rendering / simple column

File: packages/renderer/src/lib/deployments/DeploymentColumnNamespace.svelte
Comment (benoitf) [reply packages/renderer/src/lib/deployments/DeploymentColumnNamespace.svelte]: as PR has been merged, now you can use the custom rendering and drop some .svelte files

File: packages/renderer/src/lib/deployments/DeploymentColumnStatus.svelte
Comment (benoitf) [inline packages/renderer/src/lib/deployments/DeploymentColumnStatus.svelte:9]: missing a component test

File: packages/renderer/src/lib/deployments/DeploymentEmptyScreen.svelte
Comment (benoitf) [inline packages/renderer/src/lib/deployments/DeploymentEmptyScreen.svelte:6]: missing a component test

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (benoitf) [inline packages/renderer/src/lib/deployments/deployment-utils.ts]: I'm wondering why we use the toString and do not send the `.created` field directly

humanizeAge should be smarter

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (benoitf) [inline packages/renderer/src/lib/deployments/deployment-utils.ts]: here I would not use `.toString`, send `.creationTImestamp`

File: packages/renderer/src/lib/deployments/DeploymentColumnAge.svelte
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/DeploymentColumnAge.svelte]: Rebased on main to pick this up, and replaced Age and Namespace columns.

File: packages/renderer/src/lib/deployments/DeploymentColumnStatus.svelte
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/DeploymentColumnStatus.svelte:9]: Added tests for all columns.

File: packages/renderer/src/lib/deployments/DeploymentEmptyScreen.svelte
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/DeploymentEmptyScreen.svelte:6]: Added tests for all columns.

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/deployment-utils.ts]: Yes, this was poor. Updated humanizeAge to accept Date.

File: packages/renderer/src/lib/deployments/deployment-utils.ts
Comment (deboer-tim) [reply packages/renderer/src/lib/deployments/deployment-utils.ts]: Fixed.

**Comment by deboer-tim** [comment]
Rebased on main and responded to feedback. Added default name sorting and replaced Age/Namespace columns now that there are new options. Added tests, fixed humanizeAge(), and made the Age column sortable.

As per my previous comment, I felt like this was a reasonable place to merge. In a future PR I'd make the Pods column into dots and allow sorting there. Please re-review or let me know if you disagree. Thanks.

**Review by cdrage — APPROVED** [review]
LGTM! Amazing. Tested against a 2 node cluster and I was able to view deployments, sort correctly, have it update, etc. Didn't notice any issues with the UI or code.

Obviously we should merge after 1.6 / finalization of Icon?

**Comment by deboer-tim** [comment]
The deployment icon is finalized (unless we get feedback in future), it's just the Ingress/Route icon which Conor was working on (and he pinged me a final one overnight).

Yes, given the size I would expect to wait for another approval + 1.6.

**Comment by deboer-tim** [comment]
@jeffmaury or @benoitf ?

**Review by benoitf — APPROVED** [review]
I thought it was waiting the Informer stuff

**Comment by deboer-tim** [comment]
A deployment store with a regular Kube watcher (same as we do for pods) was merged before Luca came up with the informer idea. So it works fine, but I've started a separate PR to refactor deployment store to informer.

**Pull request merged**
